File: python/shapes/ishape_adapter.py
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Tuple


@dataclass
class Cell:
    face: int
    u: int
    v: int


class IShapeAdapter(ABC):
    """Interface for projecting grid coordinates to 3D space."""

    @abstractmethod
    def get_grid_size(self) -> Tuple[int, int]:
        """Return grid dimensions (u, v)."""

    @abstractmethod
    def get_face_count(self) -> int:
        """Return number of faces."""

    @abstractmethod
    def to_world(self, cell: Cell) -> Tuple[float, float, float]:
        """Convert cell to world coordinates."""

    @abstractmethod
    def wrap(self, cell: Cell, direction: str) -> Cell:
        """Wrap cell when moving off the grid in a direction."""

File: src/shapes/IShapeAdapter.ts
import * as THREE from 'three';

import type { Cell, Direction } from '../core/Grid';

export interface IShapeAdapter {
  /**
   * Return the grid dimensions in cells used by this shape.
   */
  getGridSize(): { u: number; v: number };
  /**
   * Number of distinct faces this adapter exposes. A sphere uses one,
   * while a cube uses six.
   */
  getFaceCount(): number;
  /**
   * Convert a logical grid cell to a world-space position on the shape.
   */
  toWorld(cell: Cell): THREE.Vector3;
  /**
   * Wrap a cell that moves beyond the shape's bounds to its new location
   * and face. Direction is provided for adapters that need it.
   */
  wrap(cell: Cell, dir: Direction): Cell;
}
